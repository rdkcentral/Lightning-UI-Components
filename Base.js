import lng from '@lightningjs/core';
import { withUpdates, withTags, withHandleKey, withLayout } from './mixins';
import { isComponentOnScreen } from './utils';

class Base extends lng.Component {
  static get __componentName() {
    throw new Error('Every component must have a __componentName getter');
  }

  _construct() {
    this.constructor.__componentName; // Check that __componentName is set
  }

  _update() {}

  _focus() {
    if (this._smooth === undefined) this._smooth = true;
    this._update();
  }

  _unfocus() {
    this._update();
  }

  get disabled() {
    return this._disabled;
  }

  set disabled(disabled) {
    if (this._disabled === disabled) {
      return;
    }
    this._disabled = disabled;
    this._requestUpdateDebounce();
  }

  // keep announce methods out of the update lifecycle (do not put in properties array)
  // announce methods do not need to re-render component
  get announce() {
    return this._announce;
  }

  set announce(announce) {
    this._announce = announce;
  }

  get announceContext() {
    return this._announceContext;
  }

  set announceContext(announce) {
    this._announceContext = announce;
  }

  isFullyOnScreen() {
    return isComponentOnScreen(this);
  }
}

function withMixins(baseComponent) {
  return withLayout(withUpdates(withTags(withHandleKey(baseComponent))));
}

export default withMixins(Base);
