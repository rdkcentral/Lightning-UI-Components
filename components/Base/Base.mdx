import { Canvas, Story } from '@storybook/addon-docs/blocks';
import Base from '.';

# Base

This component is used to provide a solid starting point for component development. It includes default mixins and some class methods.

## Usage


```js
import { Base } from '@lightningjs/ui-components';
class MyNewComponent extends Base {
  static _template() {
    return {
      // ...
    };
  }
};
```

## API

### Styles

Styled components can be adjusted by using [withStyles](?path=/docs/mixins-withstyles--basic)