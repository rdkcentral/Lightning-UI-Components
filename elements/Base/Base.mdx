<!-- 
  Copyright 2020 Comcast Cable Communications Management, LLC

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.

  SPDX-License-Identifier: Apache-2.0 
-->
 
import { Canvas, Story } from '@storybook/addon-docs/blocks';
import Base from '.';

# Base

This component is used to provide a solid starting point for component development. It includes default mixins and some class methods.

#### Mixins used in Base component:

mixin|description
--|--
withStyles|provides base styles
withHandleKey|processes and propogates key events
withTags|adds getters and setters for all values in `tags` property on compoent
withUpdates|adds getters and setters for all values int `properties` property on component and triggers updates when changes to any of those values
withLayout|provides `itemLayout` property to define component's width and height based on screen dimensions

## Usage


```js
import { Base } from '@lightningjs/ui-components';
class MyNewComponent extends Base {
  static _template() {
    return {
      // ...
    };
  }
};
```

## API

### Styles

Styled components can be adjusted by using [withStyles](?path=/docs/mixins-withstyles--basic)