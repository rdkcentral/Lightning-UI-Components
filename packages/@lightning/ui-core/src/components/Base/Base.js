import lng from '@lightningjs/core';
import { context } from '../../globals/index.js';
import {
  withUpdates,
  withTags,
  withHandleKey,
  withLayout,
  withThemeStyles
} from '../../mixins/index.js';
import { isComponentOnScreen } from '../../utils/index.js';

class Base extends lng.Component {
  _construct() {
    this.constructor.__componentName; // Check that __componentName is set
    if (!this.loaded) this.loaded = Promise.resolve(); // By default this is a resolved promise. Components can use _resetLoadedPromise if they requre the functionality
  }

  _init() {
    this.queueRequestUpdate();
  }

  _resetLoadedPromise() {
    this.loaded = new Promise((resolve, reject) => {
      this._resolveLoadedPromise = resolve;
      this._rejectLoadedPromise = reject;
    });
  }

  _update() {}

  _focus() {
    if (this.shouldSmooth === undefined) this.shouldSmooth = true;
    this.queueRequestUpdate();
  }

  _unfocus() {
    this.queueRequestUpdate();
  }

  applySmooth(ref, patch, smooth) {
    if (this.shouldSmooth) {
      ref.smooth = smooth || patch;
    } else {
      ref.patch(patch);
    }
  }

  // keep announce methods out of the update lifecycle (do not put in properties array)
  // announce methods do not need to re-render component
  get announce() {
    return this._announce;
  }

  set announce(announce) {
    this._announce = announce;
  }

  get announceContext() {
    return this._announceContext;
  }

  set announceContext(announce) {
    this._announceContext = announce;
  }

  get shouldSmooth() {
    return this._shouldSmooth;
  }

  set shouldSmooth(shouldSmooth) {
    this._shouldSmooth = shouldSmooth;
  }

  get _isDisabledMode() {
    return this.mode === 'disabled';
  }

  get _isUnfocusedMode() {
    return this.mode === 'unfocused';
  }

  get _isFocusedMode() {
    return this.mode === 'focused';
  }

  isFullyOnScreen() {
    return isComponentOnScreen(this);
  }

  getFocusScale() {
    return context.theme.layout.focusScale;
  }

  getUnfocusScale() {
    return 1;
  }
}

function withMixins(baseComponent) {
  return withLayout(
    withThemeStyles(withUpdates(withTags(withHandleKey(baseComponent))))
  );
}

export default withMixins(Base);
