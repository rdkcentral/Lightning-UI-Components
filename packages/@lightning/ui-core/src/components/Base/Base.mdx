import { Canvas, Story } from '@storybook/addon-docs';
import { Theme } from '../../.storybook/utils';
import Base from '.';

# Base

This component is used to provide a solid starting point for component development. It includes our default mixins and some class methods.

## Source

https://github.comcast.com/Lightning/lightning-ui/blob/develop/packages/@lightning/ui-core/src/components/Base/Base.js

## Usage

Explain how `Base` works with examples:

```js
import Base from '../../elements/Base';

class MyNewComponent extends Base {
  static _template() {
    return {
      // ...
    };
  }
}
```

## API

### Properties

### Methods

#### methodName(args): returnValue

Method description
