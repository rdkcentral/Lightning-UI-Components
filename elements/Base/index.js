import lng from '@lightningjs/core';
import {
  withStyles,
  withUpdates,
  withTags,
  withHandleKey,
  withLayout
  // withVibrant
} from '../../mixins';
import { isComponentOnScreen } from '../../utils';

const baseStyles = theme => ({
  getUnfocusScale: () => 1,
  getFocusScale: theme.getFocusScale
  // withVibrant: true
});

class Base extends lng.Component {
  _construct() {
    this._whenEnabled = new Promise(
      resolve => (this._whenEnabledResolver = resolve)
    );
    this._getFocusScale = this.styles.getFocusScale || function () {};
    this._getUnfocusScale = this.styles.getUnfocusScale || function () {};
  }

  _firstEnable() {
    this._whenEnabledResolver();
  }

  _init() {
    this.queueRequestUpdate();
  }

  _update() {}

  _focus() {
    if (this._smooth === undefined) this._smooth = true;
    this.queueRequestUpdate();
  }

  _unfocus() {
    this.queueRequestUpdate();
  }

  // keep announce methods out of the update lifecycle (do not put in properties array)
  // announce methods do not need to re-render component
  get announce() {
    return this._announce;
  }
  set announce(announce) {
    this._announce = announce;
  }

  isFullyOnScreen() {
    return isComponentOnScreen(this);
  }
}

function withMixins(baseComponent) {
  // if (baseComponent.styles.withVibrant) {
  //   baseComponent = withVibrant(baseComponent);
  // }
  return withLayout(withUpdates(withTags(withHandleKey(baseComponent))));
}

export default withMixins(withStyles(Base, baseStyles));
