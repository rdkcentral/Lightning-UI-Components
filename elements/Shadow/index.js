import lng from '@lightningjs/core';
import Base from '../../Base';
import * as styles from './Shadow.styles.js';
import { withExtensions } from '../../mixins';

class Shadow extends Base {
  static get __componentName() {
    return 'Shadow';
  }

  static get __themeStyle() {
    return styles;
  }
  static get properties() {
    return ['maskShadow'];
  }
  static get tags() {
    return ['Frame', { name: 'Shadow', path: 'Frame.Shadow' }];
  }

  _init() {
    super._init();
    this.shouldSmooth = true;
  }

  _update() {
    // To behave like Figma's shadow settings, blur value is added to width and height to expand shadow size
    const margin = this.style.spread + this.style.blur * 2; // size increase based on shadow params
    const holepunchError = 4; // extra value to tighten holepunch area so you don't see sliver of extra space

    const frameUpdate = {
      shadowMargin: margin, // store this avlue to check btwn renders
      w: this.w + margin * 2,
      h: this.h + margin * 2,
      x: this.w / 2 + this.style.offsetX,
      y: this.h / 2 + this.style.offsetY,
      mount: 0.5,

      // maskShadow prop enables a holepunch shader to mask out the parent component shape
      rtt: this.maskShadow,
      shader: this.maskShadow
        ? {
            type: lng.shaders.Hole,
            w: this.w - holepunchError,
            h: this.h - holepunchError,
            x: margin + holepunchError / 2 - this.style.offsetX,
            y: margin + holepunchError / 2 - this.style.offsetY,
            radius: this.style.radius
          }
        : undefined
    };

    const shadowPatch = {
      alpha: this.style.alpha,
      color: this.style.color
    };

    const texturePatch = {
      texture: lng.Tools.getShadowRect(
        // Underlying getShadowRect function adds blur to the canvas size, so we don't need to add it like above
        this.w + this.style.spread * 2,
        this.h + this.style.spread * 2,
        this.style.radius,
        this.style.blur
      )
    };

    if (
      !this._Frame ||
      !this.shouldSmooth ||
      // if shadow sizing has changed, patch instead of smooth
      frameUpdate.shadowMargin !== this._Frame.shadowMargin
    ) {
      this.patch({
        Frame: {
          ...frameUpdate,
          Shadow: {
            ...shadowPatch,
            ...texturePatch
          }
        }
      });
    } else {
      // Some values we don't smooth
      this._Frame.patch({
        shadowMargin: frameUpdate.shadowMargin, //store previous shadow sizing calculation
        w: frameUpdate.w,
        h: frameUpdate.h,
        rtt: frameUpdate.rtt,
        shader:
          // add shader if it's newly needed (switch from maskShadow = false to true)
          !this._Frame.shader && frameUpdate.shader
            ? frameUpdate.shader
            : // Remove shader if it's not needed (switch from maskShadow = true to false)
            !frameUpdate.shader
            ? undefined
            : // otherwise re-use last value
              this._Frame.shader,
        Shadow: { ...texturePatch }
      });

      shadowPatch.alpha = [shadowPatch.alpha, this.style.animation];
      this._Shadow.smooth = shadowPatch;

      this._Frame.setSmooth('x', frameUpdate.x, this.style.animation);
      this._Frame.setSmooth('y', frameUpdate.y, this.style.animation);

      if (this.maskShadow) {
        this._Frame.setSmooth(
          'shader.x',
          frameUpdate.shader.x,
          this.style.animation
        );
        this._Frame.setSmooth(
          'shader.y',
          frameUpdate.shader.y,
          this.style.animation
        );
        this._Frame.setSmooth(
          'shader.radius',
          frameUpdate.shader.radius,
          this.style.animation
        );
        this._Frame.setSmooth('shader.w', frameUpdate.shader.w, {
          duration: 0
        });
        this._Frame.setSmooth('shader.h', frameUpdate.shader.h, {
          duration: 0
        });
      }
    }
  }
}

export default withExtensions(Shadow);
