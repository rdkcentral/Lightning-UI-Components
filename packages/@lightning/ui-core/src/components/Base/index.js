export { default as default } from './Base.js';
