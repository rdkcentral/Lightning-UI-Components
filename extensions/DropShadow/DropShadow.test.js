import TestUtils from '../../test/lightning-test-utils';
import dropShadowExtension, {
  dropShadowExtensionGenerator
} from './DropShadow.extension';
import Base from '../../Base';
import { base as baseStyles } from '../../elements/Shadow/Shadow.styles';
import XfinityTheme from '../../themes/xfinity';

const width = 100;
const height = 150;

const baseStyleVals = baseStyles(XfinityTheme);

const generatedExtension = dropShadowExtensionGenerator({
  componentsToMask: ['ToMaskExample']
});

class Example extends Base {
  static get __componentName() {
    return 'Example';
  }
  static _template() {
    return {
      w: width,
      h: height,
      rect: true
    };
  }
}

class ToMaskExample extends Base {
  static get __componentName() {
    return 'ToMaskExample';
  }
  static _template() {
    return {
      w: width,
      h: height,
      rect: true
    };
  }
}

const createComponent = TestUtils.makeCreateComponent(
  dropShadowExtension(Example)
);

const createComponentGeneratedNoMask = TestUtils.makeCreateComponent(
  generatedExtension(Example)
);

const createComponentGeneratedWithMask = TestUtils.makeCreateComponent(
  generatedExtension(ToMaskExample)
);

describe('DropShadow Extension', () => {
  let component;
  let testRenderer;

  beforeEach(() => {
    [component, testRenderer] = createComponent();
  });

  afterEach(() => {
    component = null;
    testRenderer = null;
  });

  it('renders', () => {
    const tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with generator', () => {
    [component, testRenderer] = createComponentGeneratedNoMask();
    const tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('adds adds a DropShadow element to the component', () => {
    expect(component._DropShadow).toBeDefined();
  });

  it('Does not affect the size of the base component', () => {
    expect(component.w).toEqual(width);
    expect(component.h).toEqual(height);
  });

  it('Sizes the underlying shadow to the base component size', () => {
    expect(component._DropShadow._Frame.w).toEqual(
      component.w + 2 * (baseStyleVals.spread + baseStyleVals.blur * 2)
    );
    expect(component._DropShadow._Frame.h).toEqual(
      component.h + 2 * (baseStyleVals.spread + baseStyleVals.blur * 2)
    );
  });

  it('Applies Masked shadow to the specified component', () => {
    [component, testRenderer] = createComponentGeneratedWithMask();
    expect(component._DropShadow.maskShadow).toBe(true);
  });

  it('Does not apply a mask to the unspecified component', () => {
    [component, testRenderer] = createComponentGeneratedNoMask();
    expect(component._DropShadow.maskShadow).toBeFalsy();
  });
});
