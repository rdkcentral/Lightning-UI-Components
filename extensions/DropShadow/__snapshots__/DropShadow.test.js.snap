// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropShadow Extension renders 1`] = `
Object {
  "Component": Object {
    "DropShadow": Object {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": Object {
        "Frame": Object {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "children": Object {
            "Shadow": Object {
              "active": false,
              "alpha": 0,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": "99000000",
              "enabled": false,
              "flex": false,
              "flexItem": false,
              "h": 0,
              "isComponent": undefined,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Shadow",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": undefined,
              "tag": [Function],
              "tags": Array [
                "Shadow",
              ],
              "texture": Object {
                "type": "StaticCanvasTexture",
              },
              "visible": true,
              "w": 0,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 262,
          "isComponent": undefined,
          "mount": 0.5,
          "mountX": 0.5,
          "mountY": 0.5,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Frame",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": Array [
            "Frame",
          ],
          "visible": true,
          "w": 212,
          "x": 50,
          "y": 91,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 150,
      "hasFinalFocus": false,
      "hasFocus": false,
      "isComponent": true,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "DropShadow",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": "",
      "tag": [Function],
      "tags": Array [
        "DropShadow",
      ],
      "type": "Shadow",
      "visible": true,
      "w": 100,
      "x": 0,
      "y": 0,
      "zIndex": -1,
    },
  },
}
`;

exports[`DropShadow Extension renders with generator 1`] = `
Object {
  "Component": Object {
    "DropShadow": Object {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": Object {
        "Frame": Object {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "children": Object {
            "Shadow": Object {
              "active": false,
              "alpha": 0,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": "99000000",
              "enabled": false,
              "flex": false,
              "flexItem": false,
              "h": 0,
              "isComponent": undefined,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Shadow",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": undefined,
              "tag": [Function],
              "tags": Array [
                "Shadow",
              ],
              "texture": Object {
                "type": "StaticCanvasTexture",
              },
              "visible": true,
              "w": 0,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": false,
          "h": 262,
          "isComponent": undefined,
          "mount": 0.5,
          "mountX": 0.5,
          "mountY": 0.5,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Frame",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": Array [
            "Frame",
          ],
          "visible": true,
          "w": 212,
          "x": 50,
          "y": 91,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 150,
      "hasFinalFocus": false,
      "hasFocus": false,
      "isComponent": true,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "DropShadow",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": "",
      "tag": [Function],
      "tags": Array [
        "DropShadow",
      ],
      "type": "Shadow",
      "visible": true,
      "w": 100,
      "x": 0,
      "y": 0,
      "zIndex": -1,
    },
  },
}
`;
